custbaud_osx.c

```diff
@@ -0,0 +1,137 @@
+ * USA
+ */
+
+/***************************************************************************/
+/* Known issues:
+ * - FT232H 12MBaud not working
+ *   - using OSX El Capitan 10.11.6, FTDIUSBSerialDriver_v2_3.dmg
+ *   - tried with 2 different chips (FT232H and FT2232H)
+ *   Testing with Ftdi FT232H, which is capable to use up to 12MBaud, only
+ *   line speed up to 3MBaud were accepted. For higher baudrates we earn
+ *   a failure in the ioctl(IOSSIOSPEED) call.
+ *   But as `python -m serial.tools.miniterm` shows the same behaviour, it
+ *   looks that this is a bug or limitation in OSX and/or Ftdi driver.
+ *   Trying with PL2303 (driver version PL2303_MacOSX_1.6.1_20160309.zip),
+ *   baudrates up to 6MBaud were accepted.
+ */
+
+#if defined(__APPLE__) && defined(USE_CUSTOM_BAUD)
+
+#include "custbaud_osx.h"
+#include <string.h>
+#include <stdio.h>
+#include <errno.h>
+#include <sys/ioctl.h>
+#include <IOKit/serial/ioss.h>
+#include "term.h"
+
+/***************************************************************************/
+/* Need to undef tcsetattr to get access to the original tcsetattr()
+ * function inside our module.
+ */
+#undef tcsetattr
+
+/***************************************************************************/
+/* As we can see in OSX termios.h all the baudrate constants are transparent,
+ * like B115200=115200. There is no need for any integer <-> code translation.
+ * So we can pass any baudrate we want directly cfsetospeed() & co.
+ */
+int cfsetospeed_custom(struct termios *tiop, int speed) {
+    return cfsetospeed(tiop, speed);
+}
+
+int cfsetispeed_custom(struct termios *tiop, int speed) {
+    return cfsetispeed(tiop, speed);
+}
+
+int cfgetospeed_custom(struct termios *tiop) {
+    return cfgetospeed(tiop);
+}
+
+int cfgetispeed_custom(struct termios *tiop) {
+    return cfgetispeed(tiop);
+}
+
+/***************************************************************************/
+/* The strategy of picocom's terminal handling library is to hold all the
+ * terminal settings (including baudrate) using termios struct.
+ * Problem on OSX is, that tcsetattr() will fail if termios contains an
+ * unusual baudrate (like e.g. 12345 of 12M), The official OSX way to apply
+ * those baudrates is to use ioctl(IOSSIOSPEED) instead.
+ * Our workaround strategy is:
+ * - set the baudrate stored in termios back to a standard value (e.g. 9600)
+ * - call tcsetattr() to apply all the rest termios data to the fd
+ * - and then applying the real desired baudrate to the fd by calling ioctl(IOSSIOSPEED)
+ * Note, that in case of failed ioctl(IOSSIOSPEED), our 9600 staying
+ * configured at the fd.
+ */
+int tcsetattr_custom(int fd, int optional_actions, const struct termios *tiop) {
+    int r;
+    int workaround = 0;
+    int baudrate;
+    struct termios tios = *tiop;
+    struct termios tio0;
+    int baudrate0;
+
+    if ( fd >= 3 ) { /* don't apply this workaround for stdin/stdout/stderr */
```

Comment:
On startup (e.g. `picocom /dev/tty.usbserial -b12345`), I'd observed 2 calls to `tcsetattr():`
- `tcsetattr(3, 2, 12345)`
- `tcsetattr(0, 2, 9600)`

For the 2nd call (fd=0=stdin), when it runs into my workaround code, the `ioctl(IOSSIOSPEED)` call fail with "Inappropriate ioctl for device".

Normally, this should not happen, because I also check for `term_baud_std()` and I see 9600 baud in this case. But I dont want to rely on, that I always see 9600 baud for the stdin call.

So I decide to don't apply the workaround, if I was called with fd=0 (stdin), fd=1 (stdout) or fd=2 (stderr).